google-cloud-logging/src/main/java/com/google/cloud/logging/MonitoredResourceUtil.java

```diff
@@ -144,135 +206,152 @@ public static MonitoredResource getResource(String projectId, String resourceTyp
-  private static Resource getAutoDetectedResourceType() {
-    if (System.getenv("K_SERVICE") != null
-        && System.getenv("K_REVISION") != null
-        && System.getenv("K_CONFIGURATION") != null
-        && System.getenv("KUBERNETES_SERVICE_HOST") == null) {
-      return Resource.CloudRun;
-    }
-    if (System.getenv("GAE_INSTANCE") != null) {
-      return Resource.GaeAppFlex;
-    }
-    if (System.getenv("KUBERNETES_SERVICE_HOST") != null) {
-      return Resource.K8sContainer;
-    }
-    if (ServiceOptions.getAppEngineAppId() != null) {
-      return Resource.GaeAppStandard;
-    }
-    if (MetadataConfig.getInstanceId() != null) {
-      return Resource.GceInstance;
+  /**
+   * Heuristic to discover the namespace name of the current environment. There is no determenistic
+   * way to discover the namespace name of the process. The name is read from the {@link
+   * K8S_POD_NAMESPACE_PATH} when available or read from a user defined environment variable
+   * "NAMESPACE_NAME"
+   *
+   * @return Namespace name or empty string if the name could not be discovered
+   */
+  private static String getK8sNamespace() {
+    String value = "";
+    try {
+      value =
+          new String(Files.readAllBytes(Paths.get(K8S_POD_NAMESPACE_PATH)), StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      // if SA token is not shared the info about namespace is unavailable
+      // allow users to define the namespace name explicitly
+      value = getter.getEnv("NAMESPACE_NAME");
+      if (value == null) {
+        value = "";
+      }
     }
-    // default Resource type
-    return Resource.Global;
-  }
-
-  private static String getAppEngineModuleId() {
-    return System.getenv("GAE_SERVICE");
+    return value;
   }
 
-  private static String getAppEngineVersionId() {
-    return System.getenv("GAE_VERSION");
+  /**
+   * Distinguish between Standard and Flexible GAE environments. There is no indicator of the
+   * environment. The path to the startup-script in the metadata attribute was selected as one of
+   * the new values that explitly mentioning "flex" and cannot be altered by user (e.g. environment
+   * variable). The method assumes that the resource type is already identified as {@link
+   * Resource.AppEngine}.
+   *
+   * @return "flex" {@link String} for the Flexible environment and "standard" for the Standard.
+   */
+  private static String getAppEngineEnvironment() {
+    String value = getter.getAttribute("instance/attributes/startup-script");
+    if (value == "/var/lib/flex/startup_script.sh") {
+      return FLEX_ENV;
+    }
+    return STD_ENV;
   }
 
-  private static String getAppEngineInstanceName() {
-    return System.getenv("GAE_INSTANCE");
+  /**
+   * Retrieves a region from the qualified region of 'projects/[PROJECT_NUMBER]/regions/[REGION]'
+   *
+   * @return region string id
+   */
+  private static String getRegion() {
+    String loc = getter.getAttribute("instance/region");
```

Comment:
if these strings are used in getter as well, maybe worth extracting them to constants class and using the same string here and in getter

Reply:
Such a change creates two problems:
1) Overload the getter class with many functions that essentially do the same thing.
2) There are no much use for either of these calls. Almost any attribute is read 1-2 times during the flow, so even caching it does not make much sense.

Reply:
I didn't mean to create more methods exposing it. I meant creating constants so all strings are in one place:

```java
getter.getAttribute(ResourceStrings.INSTANCE_REGION);
```

```java
class ResourceStrings {
   public const String INSTANCE_REGION = "instance/region";
   ...
   // other similar constants
}
```